cpp/src/arrow/python/extension_type.cc

```diff
@@ -115,14 +119,21 @@ bool PyExtensionType::ExtensionEquals(const ExtensionType& other) const {
 
 std::shared_ptr<Array> PyExtensionType::MakeArray(std::shared_ptr<ArrayData> data) const {
   DCHECK_EQ(data->type->id(), Type::EXTENSION);
-  DCHECK_EQ(kExtensionName,
-            checked_cast<const ExtensionType&>(*data->type).extension_name());
+  // DCHECK_EQ(kExtensionName,
+  //           checked_cast<const ExtensionType&>(*data->type).extension_name());
   return std::make_shared<ExtensionArray>(data);
 }
 
 std::string PyExtensionType::Serialize() const {
+  PyAcquireGIL lock;
```

Comment:
Is there a particular reason you changed this implementation?

Reply:
> Is there a particular reason you changed this implementation?

I wanted to answer:

> Yes, because now the "stored" `serialized_` might not be up to date (eg in case of parametrized extension type, it is registered with a certain parametrization, but the actual instance can have a different parametrization). So you actually need to call the serialize method of the instance, instead of returning the stored `serialized_`. 

But now I think about it, I don't know anymore why I added this .. Because the `serialized_` is still set in `SetInstance`, which is called in the `__init__` of the python ExtensionType. So for a given instance, the stored `serialized_` is always correct. 

Unless your python ExtensionType is mutable (eg with the PeriodType example, that you set the `freq` parameter to a different value after creating the type instance). But I don't think we want to support this?  
(although there is not directly a way to prevent users doing this?)

Reply:
No, types are meant to be immutable.

Reply:
> No, types are meant to be immutable.

I should probably mention that in the docs then. 
And actually update the example. The two parametrized examples we have in the tests (and which I was planning to add to the docs as well) simply set a class attribute in the `__init__`, which means this attribute is mutable. So for a good example, I should probably turn them into a property (and store the actual attribute in a private attribute)